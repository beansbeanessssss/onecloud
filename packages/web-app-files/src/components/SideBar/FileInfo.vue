<template>
  <div class="file_info">
    <oc-icon :name="file.icon" size="large" class="file_info__icon" />
    <div class="file_info__body oc-text-overflow">
      <h3 data-testid="files-info-name">
        <oc-resource-name
          :name="file.name"
          :extension="file.extension"
          :type="file.type"
          :full-path="file.path"
          :is-path-displayed="false"
        />
      </h3>
      <p class="oc-my-rm">
        <template v-if="file.size > -1">{{ getResourceSize(file.size) }},</template>
        <span
          v-oc-tooltip="timeData.time"
          data-testid="files-info-mdate"
          tabindex="0"
          :aria-label="timeData.ariaLabel"
          >{{ timeData.infoString }}</span
        >
      </p>
    </div>
    <oc-button
      v-if="!publicPage() && isFavoritesEnabled"
      :aria-label="
        file.starred
          ? $gettext('Click to remove this file from your favorites')
          : $gettext('Click to mark this file as favorite')
      "
      appearance="raw"
      class="file_info__favorite"
      @click.native.stop="toggleFileFavorite(file)"
    >
      <oc-icon :class="file.starred ? 'oc-star-shining' : 'oc-star-dimm'" name="star" />
    </oc-button>
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'
import Mixins from '../../mixins'
import MixinResources from '../../mixins/resources'
import MixinRoutes from '../../mixins/routes'

export default {
  name: 'FileInfo',
  mixins: [Mixins, MixinResources, MixinRoutes],
  inject: ['displayedItem'],
  props: {
    isContentDisplayed: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    ...mapGetters(['capabilities']),
    timeData() {
      const interpolate = (obj) => {
        obj.time = this.formDateFromRFC(obj.sourceTime)
        obj.timeRelative = this.formRelativeDateFromRFC(obj.sourceTime)

        obj.infoString = this.$gettextInterpolate(obj.infoString, obj)
        obj.ariaLabel = this.$gettextInterpolate(obj.ariaLabel, obj)
        return obj
      }

      if (this.isTrashbinRoute) {
        return interpolate({
          sourceTime: this.file.ddate,
          infoString: this.$pgettext('inline info about deletion date', 'deleted %{timeRelative}'),
          ariaLabel: this.$pgettext(
            'aria label for inline info about deletion date',
            'deleted %{timeRelative} (%{time})'
          )
        })
      }

      return interpolate({
        sourceTime: this.file.mdate,
        infoString: this.$pgettext(
          'inline info about last modification date',
          'modified %{timeRelative}'
        ),
        ariaLabel: this.$pgettext(
          'aria label for inline info about last modification date',
          'modified %{timeRelative} (%{time})'
        )
      })
    },
    isFavoritesEnabled() {
      return (
        this.capabilities.files &&
        this.capabilities.files.favorites &&
        this.isContentDisplayed &&
        !this.isAnySharedWithRoute &&
        !this.isTrashbinRoute
      )
    },

    file() {
      return this.displayedItem.value
    }
  },
  methods: {
    ...mapActions('Files', ['markFavorite']),
    toggleFileFavorite(file) {
      this.markFavorite({
        client: this.$client,
        file: file
      }).then(() => {
        this.$set(this.file, 'starred', !this.file.starred)
      })
    }
  }
}
</script>

<style lang="scss">
.file_info {
  display: grid;
  grid-template-columns: auto 1fr auto;
  align-items: center;
  grid-gap: 5px;

  &__body {
    text-align: left;
    font-size: 0.75rem;

    h3 {
      font-size: 0.9rem;
      font-weight: 600;
      margin: 0;
    }
  }

  &__favorite {
    .oc-star {
      display: inline-block;

      &-shining svg {
        fill: #ffba0a !important;

        path:not([fill='none']) {
          stroke: var(--oc-color-swatch-passive-default);
        }
      }
    }
  }
}
</style>
