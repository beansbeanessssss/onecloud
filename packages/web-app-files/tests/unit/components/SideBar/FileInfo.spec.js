import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import GetTextPlugin from 'vue-gettext'
import AsyncComputed from 'vue-async-computed'

import stubs from '@/tests/unit/stubs'

import FileInfo from '@files/src/components/SideBar/FileInfo.vue'

const simpleOwnFile = {
  type: 'file',
  ownerId: 'marie',
  ownerDisplayName: 'Marie',
  mdate: 'Wed, 21 Oct 2015 07:28:00 GMT',
  size: '740'
}

const simpleDeletedFile = {
  type: 'file',
  ownerId: 'bob',
  ownerDisplayName: 'Bob',
  ddate: 'Wed, 21 Oct 2015 09:29:00 GMT'
}

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(AsyncComputed)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const selectors = {
  name: '[data-testid="files-info-name"]',
  mdate: '[data-testid="files-info-mdate"]'
}

describe('FileInfo', () => {
  it('shows file info', () => {
    const tooltipStub = jest.fn()
    const wrapper = createWrapper(simpleOwnFile, tooltipStub)
    expect(wrapper.find(selectors.name).exists()).toBeTruthy()
    expect(wrapper.find(selectors.mdate).exists()).toBeTruthy()
  })

  it('shows modification date info', () => {
    const tooltipStub = jest.fn()
    const wrapper = createWrapper(simpleOwnFile, tooltipStub)
    expect(tooltipStub).toHaveBeenCalledTimes(1)
    expect(wrapper).toMatchSnapshot()
  })

  it('shows deletion date info', () => {
    const tooltipStub = jest.fn()
    const wrapper = createWrapper(simpleDeletedFile, tooltipStub, 'files-trashbin')
    expect(tooltipStub).toHaveBeenCalledTimes(1)
    expect(wrapper).toMatchSnapshot()
  })
})

function createWrapper(testResource, tooltipStub, routeName) {
  return shallowMount(FileInfo, {
    store: new Vuex.Store({
      getters: {
        user: function () {
          return { id: 'marie' }
        },
        capabilities: jest.fn(() => ({}))
      },
      modules: {
        Files: {
          namespaced: true,
          getters: {
            highlightedFile: function () {
              return testResource
            }
          }
        }
      }
    }),
    localVue,
    stubs: stubs,
    directives: {
      OcTooltip: tooltipStub
    },
    mocks: {
      $route: {
        name: routeName || 'some-route',
        query: { page: 1 }
      },
      publicPage: () => false
    },
    methods: {
      formDateFromRFC: () => 'ABSOLUTE_TIME',
      formRelativeDateFromRFC: () => 'RELATIVE_TIME'
    },
    provide: {
      displayedItem: {
        value: testResource
      }
    }
  })
}
