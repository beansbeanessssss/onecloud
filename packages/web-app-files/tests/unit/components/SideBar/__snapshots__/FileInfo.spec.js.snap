// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInfo shows deletion date info 1`] = `
<div class="file_info">
  <oc-icon-stub size="large" class="file_info__icon"></oc-icon-stub>
  <div class="file_info__body oc-text-overflow">
    <h3 data-testid="files-info-name">
      <oc-resource-name type="file"></oc-resource-name>
    </h3>
    <p class="oc-my-rm">
      <!----> <span data-testid="files-info-mdate" tabindex="0" aria-label="deleted RELATIVE_TIME (ABSOLUTE_TIME)">deleted RELATIVE_TIME</span>
    </p>
  </div>
  <!---->
</div>
`;

exports[`FileInfo shows modification date info 1`] = `
<div class="file_info">
  <oc-icon-stub size="large" class="file_info__icon"></oc-icon-stub>
  <div class="file_info__body oc-text-overflow">
    <h3 data-testid="files-info-name">
      <oc-resource-name type="file"></oc-resource-name>
    </h3>
    <p class="oc-my-rm">740 B, <span data-testid="files-info-mdate" tabindex="0" aria-label="modified RELATIVE_TIME (ABSOLUTE_TIME)">modified RELATIVE_TIME</span></p>
  </div>
  <!---->
</div>
`;
